server/redeemer.go

```diff
@@ -0,0 +1,383 @@
+	SubscribeMaxFloatChange(sender ethcommon.Address, sink chan<- struct{}) event.Subscription
+}
+
+type Redeemer struct {
+	server    *grpc.Server
+	recipient ethcommon.Address
+	eth       eth.LivepeerEthClient
+	sm        localSenderMonitor
+	quit      chan struct{}
+}
+
+// NewRedeemer creates a new ticket redemption service instance
+func NewRedeemer(recipient ethcommon.Address, eth eth.LivepeerEthClient, sm *pm.LocalSenderMonitor) (*Redeemer, error) {
+
+	if recipient == (ethcommon.Address{}) {
+		return nil, fmt.Errorf("must provide a recipient")
+	}
+
+	if eth == nil {
+		return nil, fmt.Errorf("must provide a LivepeerEthClient")
+	}
+
+	if sm == nil {
+		return nil, fmt.Errorf("must provide a SenderMonitor")
+	}
+
+	return &Redeemer{
+		recipient: recipient,
+		eth:       eth,
+		sm:        sm,
+		quit:      make(chan struct{}),
+	}, nil
+}
+
+// Start starts a Redeemer server
+func (r *Redeemer) Start(host *url.URL) error {
+	listener, err := gonet.Listen("tcp", host.String())
+	if err != nil {
+		return err
+	}
+	defer listener.Close()
+	if err != nil {
+		return err
+	}
+	// slice of gRPC options
+	// Here we can configure things like TLS
+	opts := []grpc.ServerOption{}
+	// var s *grpc.Server
+	s := grpc.NewServer(opts...)
+	r.server = s
+
+	net.RegisterTicketRedeemerServer(s, r)
+
+	return s.Serve(listener)
+}
+
+// Stop stops the Redeemer server
+func (r *Redeemer) Stop() {
+	close(r.quit)
+	r.server.Stop()
+}
+
+// QueueTicket adds a ticket to the ticket queue
+func (r *Redeemer) QueueTicket(ctx context.Context, ticket *net.Ticket) (*net.QueueTicketRes, error) {
+	t := pmTicket(ticket)
+	if r.recipient != t.Recipient {
+		return nil, status.Error(codes.PermissionDenied, fmt.Sprintf("invalid ticket recipient 0x%x, expected %v", ticket.TicketParams.Recipient, r.recipient.Hex()))
+	}
+	if err := r.sm.QueueTicket(t); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	glog.Infof("ticket queued sender=0x%x", ticket.Sender)
+
+	return &net.QueueTicketRes{}, nil
+}
+
+// MaxFloat is a unary RPC method to request the max float value for a sender
+func (r *Redeemer) MaxFloat(ctx context.Context, req *net.MaxFloatReq) (*net.MaxFloatUpdate, error) {
+	maxFloat, err := r.sm.MaxFloat(ethcommon.BytesToAddress(req.Sender))
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &net.MaxFloatUpdate{MaxFloat: maxFloat.Bytes()}, nil
+}
+
+// MonitorMaxFloat starts a server-side stream to the client to send max float updates for sender specified in the request
+func (r *Redeemer) MonitorMaxFloat(req *net.MaxFloatReq, stream net.TicketRedeemer_MonitorMaxFloatServer) error {
+	sender := ethcommon.BytesToAddress(req.Sender)
+
+	sink := make(chan struct{}, 10)
+	sub := r.sm.SubscribeMaxFloatChange(sender, sink)
+	defer sub.Unsubscribe()
+	for {
+		select {
+		case <-r.quit:
+			return nil
+		case err := <-sub.Err():
+			if err == nil {
+				return status.Error(codes.Canceled, "subscription closed")
```

Comment:
I believe nil on the channel means the channel and thus subscription is closed for an unexpected reason. In this case we should close the stream and inform the client to clear it's local cache so it can start a new monitoring sequence.

Reply:
According to the [docs](https://godoc.org/github.com/ethereum/go-ethereum#Subscription) for `Subscription`, when `Unsubscribe()` is called the error channel will be closed. At that point, the last value sent on the channel due to the close operation will be nil (which in this case is the zero value of the channel data type). 

Technically, we shouldn't ever receive nil on the error channel since the `Unsubscribe()` call is deferred until after the for loop returns. But, *if* there is a nil error returned on the error channel I don't think returning an error is correct since no error actually occurred.

Reply:
Ah I see I thought that the error channel only received a nil value when it was closed due to `Unsubscribe()` being called, but I see in the docs that the error channel could also be closed due to an issue with the subscription. In that case, returning an error here makes sense 👍